### ISSUE TITLE
Terraform can't remove VM Diagnositics once Enabled

### ISSUE THREAD

**Issue by rohrerb** [terraform-providers/terraform-provider-azurerm#1167]
Hello,

Steps:

1. Create a vm **without** diagnostics enabled.  Verify VM is created with no diagnostics via the portal or cli.

2. Then add storage account
```terraform
resource "azurerm_storage_account" "diag_storage_account" {
  count = "${var.instance_count}"
  name = "${format("%sstg%03diag", local.base_hostname, count.index + 1)}"
  resource_group_name = "${var.resource_group_name}"
  location = "${var.azure_location}"
  account_tier = "Standard"
  account_replication_type = "LRS"
  enable_blob_encryption = "true"
}
```

3. Add the boot diagnostics inside of the VM block
```terraform
  boot_diagnostics {
    enabled = true
    storage_uri = "${element(azurerm_storage_account.diag_storage_account.*.primary_blob_endpoint, count.index)}"
  }
```

4. Re-Run terraform apply
5. You will see the actions lists removing the boot diagnostics.
```shell
module.wtes.azurerm_virtual_machine.vm
      boot_diagnostics.#:             "1" => "0"
      boot_diagnostics.0.enabled:     "true" => "false"
      boot_diagnostics.0.storage_uri: "abcdef.blob.core.usgovcloudapi.net/" => ""
```

It looks like it updates but really doesn't.  Re-run a apply and you will see the same actions to remove the boot diagnostics. 

It is no help even if you manually remove it from the state file.

Any guidance on this would be appreciated.

Thank you,
Brandon

**Comment by VaijanathB** [comment]
This underneath issue needs to be fixed to properly address this in Terraform. https://github.com/Azure/azure-rest-api-specs/issues/2994

**Comment by bostonmoto** [comment]
Hi there folks any update on when this issue will be addressed or any potential workarounds?

**Comment by bostonmoto** [comment]
For those looking for a workaround to this issue. Here is what I did to avoid seeing this in my plan. I added is as part of my virtual_machine resource.

lifecycle {
  ignore_changes = [
  "boot_diagnostics.#",
  "boot_diagnostics.0.storage_uri"
     ]

**Comment by tombuildsstuff** [comment]
@bostonmoto this is still blocked on the upstream API issue: https://github.com/Azure/azure-rest-api-specs/issues/2994

**Comment by kenorb** [comment]
Still the issue with v0.11.14.

```
  ~ module.foobar.azurerm_virtual_machine.notary
      boot_diagnostics.#:             "1" => "0"
      boot_diagnostics.0.storage_uri: "https://tempnotary5.blob.core.windows.net" => ""
```

Despite _Boot diagnostics are not configured for this virtual machine._.

**Comment by MacFlurry** [comment]
+1 issue still present.
Any update on this issue ?

**Comment by jlact** [comment]
For those using 0.12 the equivalent workaround is:
```
lifecycle {
    ignore_changes = [
      "boot_diagnostics[0].storage_uri"
    ]
}
```

**Comment by tombuildsstuff** [comment]
👋 

To give an update here - I'm currently taking a look at the new VM & VMSS resources as a part of 2.0 (which is outlined #2807), which as a part of we're adding new resources and covering test cases for the existing bugs to ensure these are supported.

As a part of building out the new VM Scale Set resources I've added a test-case for enabling/disabling/enabling Boot Diagnostics, which works - and as such will ship as a part of that resource. Since the API's for VM Scale Set and VM's are similar, I'm assuming (but haven't yet confirmed) that this is also true for the Virtual Machine API's - once we get to the new Virtual Machine resource we'll add a test case and confirm for sure; however this'll be coming to the new Virtual Machine resources `azurerm_linux_virtual_machine` and `azurerm_windows_virtual_machine` rather than the existing `azurerm_virtual_machine` resource - since to achieve this a different schema is required than is present in the existing resource.

As such for the moment I've assigned this to the 2.0 milestone and we'll give another update in the near future when we've been able to confirm this.

Thanks!

**Comment by tombuildsstuff** [comment]
hi @rohrerb 

Taking a look through here the `azurerm_virtual_machine` resource has since been superseded by the `azurerm_linux_virtual_machine` and `azurerm_windows_virtual_machine` resources. Whilst the existing `azurerm_virtual_machine` resource continues to be available in 2.x releases of the Provider, it's feature-frozen and thus no longer receiving any bugfixes or enhancements to maintain compatibility.

As such I'm going to close this issue for the moment - but I'd recommend looking into the [`azurerm_linux_virtual_machine`](https://www.terraform.io/docs/providers/azurerm/r/linux_virtual_machine.html) and [`azurerm_windows_virtual_machine`](https://www.terraform.io/docs/providers/azurerm/r/windows_virtual_machine.html) resources going forward which contain a fix for this.

Thanks!
